### PR TITLE
Added the Windows 10 (UAP) platform to protobuild (pull from MonoGame)

### PR DESCRIPTION
**Pull request by DDReaper**
As pre request here:
https://github.com/mono/MonoGame/issues/3405#issuecomment-156738010

Created a PR to add Windows 10 (UAP) platform support to ProtoBuild .xslt's

(Still not sure about naming them UAP as the proper Windows 10 definition is UWP, but that is the current standard used in MG presently)

### CODE REVIEW

**Comment by tomspilman** [comment]
> Still not sure about naming them UAP as the proper Windows 10 definition is UWP

I'm good with changing that... we just have to fix all the MonoGame code to match.  Just not sure how disruptive that will be for users.

**Comment by DDReaper** [comment]
Maybe we should just call it the Windows10 build then, save any particular confusion?

**Comment by tomspilman** [comment]
> Maybe we should just call it the Windows10 build then

Perhaps it could be that we're being optimistic that UWP sticks around past Win10... although I suspect it will.

This is all internal to the projects and classes anyway.

**Comment by DDReaper** [comment]
Well, Win10 is meant to be the "Last" version of Windows, the UWP part was just meant to signify the "one build" for all platforms, however it's still known as Win 10 builds (or Windows Universal without the 8, which is far too confusing).
If we change MG however, we need to ensure all the templates and other deliverabled follow suit.

Ok, if Win10 it it, I'll update this PR and then work on another PR for MG to update the term.

So do you agree @tomspilman ?

**Comment by tomspilman** [comment]
We need to see what @hach-que says first.

**Comment by DDReaper** [comment]
Fair enough, @hach-que over to you?

**Comment by hach-que** [comment]
I think "WindowsUniversal" might be the best platform name.  "WindowsUAP" and "WindowsUWP" don't communicate what the platform is unless you know the acronyms, and for every other platform we've stuck with a platform name is understandable without doing that.

File: Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt [hunk 1]
```diff
@@ -121,6 +121,12 @@
             <TargetPlatformVersion>8.1</TargetPlatformVersion>
             <MinimumVisualStudioVersion>12</MinimumVisualStudioVersion>
           </xsl:when>
+          <xsl:when test="$root/Input/Generation/Platform = 'WindowsUAP'">
+			  <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
```
Comment (hach-que) [inline Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt]: Fix the indentation on these lines to use spaces

File: Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt [hunk 2]
```diff
@@ -267,6 +273,9 @@
               <xsl:when test="$root/Input/Generation/Platform = 'WindowsPhone81'">
                 <xsl:text>PLATFORM_WINDOWSPHONE81</xsl:text>
               </xsl:when>
+              <xsl:when test="$root/Input/Generation/Platform = 'WindowsUAP'">
+                <xsl:text>PLATFORM_WINDOWSUAP</xsl:text>
```
Comment (hach-que) [inline Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt]: Let's go with PLATFORM_WINDOWSUNIVERSAL here as well

File: Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt [hunk 3]
```diff
@@ -938,7 +950,10 @@
           </xsl:otherwise>
         </xsl:choose>
         <xsl:choose>
-          <xsl:when test="$root/Input/Generation/Platform = 'Windows8' or $root/Input/Generation/Platform = 'WindowsPhone81'">
+			<xsl:when test="$root/Input/Generation/Platform = 'WindowsUAP'">
```
Comment (hach-que) [inline Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt]: Indentation again - use spaces

Comment (hach-que) [inline Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt]: Also don't mix this choose statement here; this choose statement is for the `<ProductVersion>` tag.  To introduce the `<FileAlignment>` tag, use a separate `<xsl:if>` or `<xsl:choose>` block.

File: Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt [hunk 4]
```diff
@@ -1957,7 +1978,13 @@
           </PropertyGroup>
           <Import Project="$(MSBuildExtensionsPath)\Microsoft\WindowsXaml\v$(VisualStudioVersion)\Microsoft.Windows.UI.Xaml.CSharp.targets" />
         </xsl:when>
-        <xsl:when test="$root/Input/Generation/Platform = 'WindowsPhone81'">
+		  <xsl:when test="$root/Input/Generation/Platform = 'WindowsUAP'">
```
Comment (hach-que) [inline Protobuild.Internal/BuildResources/GenerateProject.CSharp.xslt]: Indentation again

**Comment by hach-que** [comment]
(the other reason for preferring WindowsUniversal over something with a version number in it is that this platform is applicable to other devices, such as XBox and HoloLens, and I don't want to communicate that it's restricted to a particular version of Windows, or only a subset of operating systems that are explicitly named Windows)

**Comment by DDReaper** [comment]
My only qualm with just WindowsUniversal I that with windows 8 we also had Universal projects, so could lead to confusion. But not against it.
Noted your line comments, will update as soon as we've agreed on a name 😁

**Comment by DDReaper** [comment]
Sorry, wrong button 😒

**Comment by hach-que** [comment]
Also FYI I'm about to introduce changes upstream that switch Mac OS over to use the Unified Xamarin API by default, because that is required for post-build hooks to work.  This will impact MonoGame upstream if updating to a later version of Protobuild.

This can be temporarily avoided by setting the `UseLegacyMacAPI` option to true, but as with the iOS Unified changes, this needs to happen eventually since I expect Apple or Xamarin will drop support for the Classic API on Mac.

**Comment by DDReaper** [comment]
So to close up this naming thing, what name are we agreeing on @tomspilman @hach-que or the win 10 universal platform?

**Comment by tomspilman** [comment]
> So to close up this naming thing, what name are we agreeing on

Still feel that "WindowsUWP" is fine... yes the acronym ... but is it any worse than PCL or WindowsGL.  Remember the audience is supposed to be programmers/developers here.

I hate how long `WindowsUniversal` is as a name and the potential confusion for the Windows 8 Universal apps... but i can live with that if nothing better is decided on.

Your call @hach-que .

**Comment by hach-que** [comment]
Considering that `WindowsGL` is MonoGame specific and is no longer an official platform in Protobuild, and `PCL` is a speciality platform that you only really use if you know what you are doing... I still feel as though `WindowsUniversal` is the best option here (it's not significantly worse than `WindowsPhone81`).

**Comment by hach-que** [comment]
(I'm also expecting that Windows 8 Universal Apps will disappear in the near future; Microsoft will almost certainly phase all of that out in favour of the new Universal app platform, especially with the intention to automatically upgrade Windows 7 and higher to 10 through Windows Update).

**Comment by DDReaper** [comment]
Ok, fair point. Enough talk and let's just get on with it and do it (😁 life of Brian nod).
Will update this pr, then when done we need another mg pr to update protobuild again (does that mean the xslt's can be dropped?)
Then I can finish up my other core pcl pr. Phew, good things always take the most time.
Also good we've been able to resync the projects again.

Then we'll need to look at @hach-que  comment about ios @tomspilman 
Last Q, is anything needed as part of this for dean's tvos platform, I've not checked?

**Comment by hach-que** [comment]
Dean's tvOS platform introduces more changes to your copies of GenerateProject, but I want that platform upstreamed so feel free to submit another PR.

The Mac OS universal change won't impact iOS (we already did this a while back when MonoGame needed to move to it), but it will impact native Mac OS applications.

**Comment by DDReaper** [comment]
Back to you @hach-que Fixed indentations (actually found a few other preceding ones and resolved them to, tabs be gone). Also named the new platform WindowsUniversal as requested.
Will sort out a separate PR for the TVOS platform so it can just be one merge back to MG

**Comment by DDReaper** [comment]
I'll wait until this PR is consumed to add the TVOS platform, else it'll get too confusing.  Should be much quicker though looking at deans pending changes

**Comment by tomspilman** [comment]
> actually found a few other preceding ones and resolved them to, tabs be gone

Note that solution .sln is very sensitive to whitespace...  one extra space in an indent and VS sees the file as incorrect and will apply its own configuration on the next save.  Just be sure that when the generated solution is opened in VS... hitting Save All doesn't cause it to overwrite the .sln file.

**Comment by DDReaper** [comment]
Thanks for the pointer Tom, I'll double check the new config with the branch I'm running

**Comment by hach-que** [comment]
Yes I should have been clearer; the GenerateSolution.xslt must have tabs because of VS requirements, while everywhere else should be intended with spaces.

**Comment by hach-que** [comment]
(Also is this ready to merge?)

**Comment by hach-que** [comment]
I have rebase-squashed these changes and built them to https://github.com/hach-que/Protobuild/tree/DDReaper-protobuild_WindowsUAP.  Let me know if this works (or if there are other changes to be made, push them here and I'll reintegrate them).

**Comment by DDReaper** [comment]
I didn't touch the generate solution xslt, just the csharp xslt for the indentations.
Just testing with mg and seems fine so far, going to submit a separate pr to rename the windowsuap to windowsuniversal to mg, as it's a little more impactful than originally anticipated.
The rename will need to be done first in mg before pushing this back.
Should be done testing today

**Comment by DDReaper** [comment]
All tests past and generate is working fine.
Also tested against a version of MG with the new WindowsUniversal platform (#4255) and checked each project loads and builds properly, no issues found.  Ready to go.

Will review dean's TVOS xslt changes next once this is merged/consumed and sumbit a second PR

**Comment by hach-que** [comment]
Oops, I thought I merged this already...

**Comment by hach-que** [comment]
Squash / rebased into master.

**Pull request closed without merging**